src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/UniversalCryptoOneShot.cs

```diff
@@ -20,6 +20,10 @@ internal static class UniversalCryptoOneShot
             if (input.Length % cipher.PaddingSizeInBytes != 0)
                 throw new CryptographicException(SR.Cryptography_PartialBlock);
 
+            bool cipherHandlesPadding = cipher.HandlesPadding;
+            if (cipherHandlesPadding)
```

Comment:
The rest of the implementation of this class is iffy once webcrypto starts getting involved since it relies on being able to perform individual blocks (`Transform` instead of `TransformFinal`), whereas webcrypto gives you a one-shot.

For the sake of simplicity, I think you could implement it like:

```C#
if (cipherHandlesPadding)
{
    cipher.ValidatePaddingMode(paddingMode);

    try
    {
        bytesWritten = cipher.TransformFinal(input, output);
        return true;
    }
    catch (Exception ex) when (ex is ArgumentException or CryptographicException)
    {
        bytesWritten = 0;
        // Don't leave partially copied data in the output
        CryptographicOperations.ZeroMemory(output);
        return false;
    }
}
```

An alternative might be to not use this class at all - and instead change `AesImplementation.Browser` to not use this one-shot mechanism at all.

Reply:
> An alternative might be to not use this class at all - and instead change AesImplementation.Browser to not use this one-shot mechanism at all.

I had started down this route, but it felt like I was "fighting the design" too much. I would need to make a bunch of .csproj changes on which .cs files were included, and the existing one would need to be named `*.NotBrowser.cs`, etc. Also, I would need to handle all the cases that `OneShotDecrypt` already handled for me. In the end, it just felt "wrong".

I wasn't sure if the `cipher.HandlesPadding` / `cipher.ValidatePaddingMode` was a great design choice either, but I felt like it fit into the existing model better.

@bartonjs @GrabYourPitchforks - any thoughts here?

Reply:
I would consider splitting `UniversalCryptoOneShot` then by implementation, then. It's fairly complicated as I've added to it over the past few months trying to eke out more performance.

I think this diff highlights what my suggestion is: https://github.com/dotnet/runtime/commit/7741c51c58d0d58507de3c5d0f3de557fa7eea08

Reply:
Ideally, `RijndaelManagedTransform.Browser.cs` could be taught how to `TransformFinal` directly to a `ReadOnlySpan<byte>` instead of following the required "return an allocation" pattern. If `DecryptData` and `EncryptData` can write to span instead of "write to a byte array or maybe make one for you if you want" would help and could get rid of some "rent, transform, copy" patterns.

Reply:
The latest changes have resolved this. PTAL